### DISCUSSION TITLE
PATH of bash shell in Windows is not correct.

### DISCUSSION THREAD

**Discussion by liudonghua123** [community/community#26149] [Actions]
<p>I have these pieces of configurations in my workflow.</p>
<pre><code>steps:
      - uses: actions/checkout@v1
      - uses: DanTup/gh-actions/setup-dart@master
      - name: Build
        run: |
          pub get
          mkdir build
          dart2native bin/main.dart -v -o build/${{ matrix.output-name }}
        shell: bash
</code></pre>
<p>I got pub: command not found issue.<br>
If I remove shell: bash and use the builtin shell, then it’s ok.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/liudonghua123">@liudonghua123</a> ,</p>
<p>You can try using " <strong>pub.bat</strong>" instead of directly " <strong>pub</strong>" to run the commands, this can also work on bash shell.</p>
<p>For example:</p>
<pre><code>- name: Run pub with bash
  shell: bash
  run: |
    pub.bat get

- name: Run pub with pwsh
  shell: pwsh
  run: |
    pub.bat get
</code></pre>

**Comment by liudonghua123** [comment]
<p>I can’t use this syntax, because I used matrix strategy and I need to run the command in MacOS and Linux too. In my opinion, we should find the true reasons, fix the PATH inheritance in bash of git in Windows. Command of Windows can run without extension defined in PATHEXT environment variable.</p>

**Comment by liudonghua123** [comment, marked as answer]
<p>Finally, I refactor the code like the following.</p>
<pre><code>- name: Build
        run: |
          # check Linux/MacOS or Windows
          if hash pub 2&gt;/dev/null; then PUB=pub; DART2NATIVE=dart2native; else PUB=pub.bat; DART2NATIVE=dart2native.bat; fi
          mkdir build
          $PUB get
          $DART2NATIVE bin/main.dart -v -o build/${{ matrix.output-name }}
        shell: bash
</code></pre>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/liudonghua123">@liudonghua123</a> ,</p>
<p>Thanks for sharing your workarond.<br>
I also tested the pub commands after installing the Dart-SDK on my local Windows machine, and I also reproduce same issue. Looks like, this problem occurs on Windows, and it should not be an issue for GitHub Actions.<br>
Similar to your shared workaround, I also have a workaround to run the pub commands with bash shell on different OSs.</p>
<pre><code>jobs:
  TEST:
    name: Test
    runs-on: ${{ matrix.os }}
    env:
      PUB: pub
    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-latest, windows-latest, macOS-latest]
    steps:
      - name: setup-dart
        uses: DanTup/gh-actions/setup-dart@master

      - name: Change env
        if: matrix.os == 'windows-latest'
        run: echo "::set-env name=PUB::pub.bat"

      - name: Test pub command
        shell: bash
        run: |
          $PUB --help
</code></pre>

**Comment by liudonghua123** [comment]
<blockquote>
<p><a class="mention" href="/brightran">@brightran</a> wrote:</p>
</blockquote>
<blockquote>
<p><a class="mention" href="/liudonghua123">@liudonghua123</a> ,</p>
</blockquote>
<blockquote>
<p>Thanks for sharing your workarond.<br>
I also tested the pub commands after installing the Dart-SDK on my local Windows machine, and I also reproduce same issue. Looks like, this problem occurs on Windows, and it should not be an issue for GitHub Actions.<br>
Similar to your shared workaround, I also have a workaround to run the pub commands with bash shell on different OSs.</p>
</blockquote>
<blockquote>
<p>jobs:<br>
TEST:<br>
name: Test<br>
runs-on: ${{ matrix.os }}<br>
env:<br>
PUB: pub<br>
strategy:<br>
fail-fast: false<br>
matrix:<br>
os: [ubuntu-latest, windows-latest, macOS-latest]<br>
steps:</p>
<ul>
<li>name: setup-dart<br>
uses: DanTup/gh-actions/setup-dart@master</li>
</ul>
</blockquote>
<blockquote>
<ul>
<li>name: Change env<br>
if: matrix.os == ‘windows-latest’<br>
run: echo “::set-env name=PUB::pub.bat”</li>
</ul>
</blockquote>
<blockquote>
<ul>
<li>name: Test pub command<br>
shell: bash<br>
run: |<br>
$PUB --help</li>
</ul>
</blockquote>
<p>Thanks, your solution is more elegance and readable.</p>
